### PR TITLE
[Processing] Fix processing refactor fields algorithm using expression context generators

### PR DESCRIPTION
**Pull request by arnaud-morvan**
- Add QgsExpressionContextGenerator inheritances in sip
- Fix processing refactor fields algorithm

### CODE REVIEW

**Comment by arnaud-morvan** [comment]
@m-kuhn @nyalldawson : could some one have a look, and merge ?

File: python/plugins/processing/algs/qgis/ui/FieldsMappingPanel.py [hunk 1]
```diff
@@ -87,20 +87,25 @@ def testExpression(self, row):
         self._errors[row] = None
         field = self._mapping[row]
         expression = QgsExpression(field['expression'])
+        context = self.contextGenerator().createExpressionContext()
+        expression.prepare(context)
```
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/ui/FieldsMappingPanel.py]: I would move that down just before the loop (check for parser errors before preparing)

**Comment by m-kuhn** [comment]
Looks good to me, thanks

**Comment by nyalldawson** [comment]
I *think* the processing specific expression variables should also be added here, but I'm not familiar enough with the processing code to know if they are appropriate in this context.

**Comment by arnaud-morvan** [comment]
I do not know exactly about processing variables, but it seems to be for expressions support in parameter input dialogs. Here expressions will be used in GeoAlgorithm.processAlgorithm. I think that layer context is adequate.
@volaya : can you confirm ?

**Comment by arnaud-morvan** [comment]
@m-kuhn : Regarding QgsExpression::isValid prepare should be done first : https://github.com/qgis/QGIS/blob/3d6e7c8bdbfa455702fc523c63992a5c45e5ad23/src/core/qgsexpression.cpp#L3497:L3503
I would use isValid, but I do not manage to get error message in python using static method isValid.

**Comment by m-kuhn** [comment]
If you are going to construct an expression anyway, I don't think there's a need to run isValid() on it first.
Just create it, check for parser errors and then prepare.

**Comment by m-kuhn** [comment]
Nevertheless, I pushed a fix for this problem to a current pull request where I'm working on the expressions anyway https://github.com/qgis/QGIS/pull/3472/commits/dfc35b61f07b14aa8fe95fc6675176769bb23923

**Comment by arnaud-morvan** [comment]
To be consistent, I've also updated the algorithm with expression context from generator.

As prepare could set the parserErrorString, I prefer to test it after prepare, like in isValid / checkExpression.

**Comment by m-kuhn** [comment]
> As prepare could set the parserErrorString

Shouldn't actually happen but no big deal to leave it :)

**Comment by m-kuhn** [comment]
Ok, good for me :)

**Comment by volaya** [comment]
@arnaud-morvan Not sure it is exactly the same, but as part of the changes in the parameters architecture, i have added a parameter evaluation before running the algorithm. See https://github.com/volaya/QGIS/commit/80227f6e8c708260525e991d4104c0a5823aba04

That allows a better use of expressions, since they are not evaluated when the value is set, but when running the algorithm, so the algorithm is called with the expression itself, not the result of its evaluation. Does that solve the same issue as this PR? IF so, maybe wait until we merge that and then work on top of it if changes are needed

**Comment by arnaud-morvan** [comment]
I do not see use case of processing context values here, but it could be added in another thread if needed.
This PR is a fix, ready for merge when Travis is happy.

**Pull request merged**